File: labs-opensource/M08-conversation-management/solution/conversation_manager.js
/**
 * M08 Lab - Steps 1+2: ConversationManager + SlidingWindowManager — SOLUTION
 * ===========================================================================
 * Run: node conversation_manager.js
 */

import OpenAI from "openai";

import { pathToFileURL } from "node:url";
const client = new OpenAI({ baseURL: "http://localhost:11434/v1", apiKey: "ollama" });

export class ConversationManager {
  constructor({ systemPrompt = "You are a helpful assistant.", model = "mistral" } = {}) {
    this.systemPrompt = systemPrompt;
    this.model = model;
    this.messages = [];
    this.totalInputTokens = 0;
    this.totalOutputTokens = 0;
  }

  addUserMessage(content) {
    this.messages.push({ role: "user", content });
  }

  addAssistantMessage(content) {
    this.messages.push({ role: "assistant", content });
  }

  getMessages() {
    return [{ role: "system", content: this.systemPrompt }, ...this.messages];
  }

  async send(userMessage) {
    this.addUserMessage(userMessage);
    try {
      const response = await client.chat.completions.create({
        model: this.model,
        messages: this.getMessages(),
      });
      if (response.usage) {
        this.totalInputTokens += response.usage.prompt_tokens;
        this.totalOutputTokens += response.usage.completion_tokens;
      }
      const assistantText = response.choices[0].message.content;
      this.addAssistantMessage(assistantText);
      return assistantText;
    } catch (error) {
      this.messages.pop(); // remove failed user message
      throw new Error(`API call failed: ${error.message}`);
    }
  }

  getTokenUsage() {
    return {
      totalInput: this.totalInputTokens,
      totalOutput: this.totalOutputTokens,
      messages: this.messages.length,
    };
  }
}

export class SlidingWindowManager extends ConversationManager {
  constructor({ windowSize = 10, ...opts } = {}) {
    super(opts);
    this.windowSize = windowSize;
  }

  getMessages() {
    let windowed = this.messages.length <= this.windowSize
      ? [...this.messages]
      : this.messages.slice(-this.windowSize);

    // History sent to the API should start with a user turn
    if (windowed[0]?.role === "assistant") {
      windowed = windowed.slice(1);
    }
    return [{ role: "system", content: this.systemPrompt }, ...windowed];
  }
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  console.log("=".repeat(50));
  console.log("TEST 1: Basic ConversationManager (full history)");
  console.log("=".repeat(50));
  const mgr = new ConversationManager({ systemPrompt: "You are a concise coding tutor. Answer in 1-2 sentences." });
  for (const q of ["What is a list comprehension in Python?", "Show me an example with filtering."]) {
    console.log(`\nQ: ${q}`);
    console.log(`A: ${(await mgr.send(q)).slice(0, 150)}`);
    console.log(`   usage so far:`, mgr.getTokenUsage());
  }

  console.log("\n" + "=".repeat(50));
  console.log("TEST 2: SlidingWindowManager (windowSize=6)");
  console.log("=".repeat(50));
  const win = new SlidingWindowManager({ windowSize: 6, systemPrompt: "You are a concise coding tutor." });
  for (const q of ["What is Python?", "What are variables?", "Explain loops.",
                   "What are functions?", "Explain classes.", "What is inheritance?"]) {
    await win.send(q);
    console.log(`Q: ${q}  (stored=${win.messages.length}, sent=${win.getMessages().length - 1})`);
  }

  console.log(`\nStored: ${win.messages.length} messages`);
  console.log(`Sent (excl. system): ${win.getMessages().length - 1} messages`);
}

File: labs-opensource/M08-conversation-management/solution/conversation_manager.py
"""
M08 Lab - Steps 1+2: ConversationManager + SlidingWindowManager — SOLUTION
===========================================================================
Run: python conversation_manager.py
"""

from dataclasses import dataclass, field

from openai import OpenAI

client = OpenAI(base_url="http://localhost:11434/v1", api_key="ollama")


@dataclass
class ConversationManager:
    """Full-history conversation manager with token tracking."""

    system_prompt: str = "You are a helpful assistant."
    model: str = "mistral"
    messages: list = field(default_factory=list)

    def __post_init__(self):
        self.total_input_tokens = 0
        self.total_output_tokens = 0

    def add_user_message(self, content: str) -> None:
        self.messages.append({"role": "user", "content": content})

    def add_assistant_message(self, content: str) -> None:
        self.messages.append({"role": "assistant", "content": content})

    def get_messages(self) -> list:
        return [{"role": "system", "content": self.system_prompt}] + list(self.messages)

    def send(self, user_message: str) -> str:
        self.add_user_message(user_message)
        try:
            response = client.chat.completions.create(
                model=self.model,
                messages=self.get_messages(),
            )
            if response.usage:
                self.total_input_tokens += response.usage.prompt_tokens
                self.total_output_tokens += response.usage.completion_tokens
            assistant_text = response.choices[0].message.content
            self.add_assistant_message(assistant_text)
            return assistant_text
        except Exception as e:
            self.messages.pop()  # remove failed user message
            raise RuntimeError(f"API call failed: {e}") from e

    def get_token_usage(self) -> dict:
        return {
            "total_input": self.total_input_tokens,
            "total_output": self.total_output_tokens,
            "messages": len(self.messages),
        }


class SlidingWindowManager(ConversationManager):
    """Stores everything, but only SENDS the most recent N messages."""

    def __init__(self, window_size: int = 10, **kwargs):
        super().__init__(**kwargs)
        self.window_size = window_size

    def get_messages(self) -> list:
        if len(self.messages) <= self.window_size:
            windowed = list(self.messages)
        else:
            windowed = self.messages[-self.window_size:]

        # History sent to the API should start with a user turn
        if windowed and windowed[0]["role"] == "assistant":
            windowed = windowed[1:]

        return [{"role": "system", "content": self.system_prompt}] + windowed


if __name__ == "__main__":
    print("=" * 50)
    print("TEST 1: Basic ConversationManager (full history)")
    print("=" * 50)
    mgr = ConversationManager(system_prompt="You are a concise coding tutor. Answer in 1-2 sentences.")
    for q in ["What is a list comprehension in Python?", "Show me an example with filtering."]:
        print(f"\nQ: {q}")
        print(f"A: {mgr.send(q)[:150]}")
        print(f"   usage so far: {mgr.get_token_usage()}")

    print("\n" + "=" * 50)
    print("TEST 2: SlidingWindowManager (window_size=6)")
    print("=" * 50)
    win = SlidingWindowManager(window_size=6, system_prompt="You are a concise coding tutor.")
    for q in ["What is Python?", "What are variables?", "Explain loops.",
              "What are functions?", "Explain classes.", "What is inheritance?"]:
        win.send(q)
        print(f"Q: {q}  (stored={len(win.messages)}, sent={len(win.get_messages()) - 1})")

    print(f"\nStored: {len(win.messages)} messages")
    print(f"Sent (excl. system): {len(win.get_messages()) - 1} messages")
